src/libsyntax/parse/parser.rs

```diff
@@ -6214,14 +6214,24 @@ impl<'a> Parser<'a> {
         })
     }
 
-    pub fn ensure_complete_parse<F>(&mut self, allow_semi: bool, on_err: F)
-        where F: FnOnce(&Parser)
+    pub fn ensure_complete_parse(&mut self, kind: ExpansionKind, macro_name: ast::Name, span: Span)
     {
-        if allow_semi && self.token == token::Semi {
+        // We allow semicolons at the end of expressions -- e.g. the semicolon in
+        // `macro_rules! m { () => { panic!(); } }` isn't parsed by `.parse_expr()`,
+        // but `m!()` is allowed in expression positions (c.f. issue #34706).
```

Comment:
If a macro author wants their macro to be usable in an expression context, couldn't they just not include the trailing semicolon. Afaik, the trailing semicolon isn't needed for the macro to be usable in a statement context.